eservice/pdo/eservice/enclave/enclave/block_store.cpp

```diff
@@ -0,0 +1,187 @@
+         * allocated data later
+         */
+        *valueSize = valueStr.size() / 2;
+        *value = (uint8_t *)malloc(*valueSize);
+        if (!*value) {
+            Log(PDO_LOG_ERROR,
+                "Failed to allocate %zu bytes for get return value.",
+                *valueSize);
+            *valueSize = 0;
+            result = PDO_ERR_MEMORY;
+            goto done;
+        }
+
+        // Deserialize the data from the cache into the buffer
+        HexStringToBinary(*value, *valueSize, valueStr);
+
+        result = PDO_SUCCESS;
+    }
+
+done:
+    pthread_spin_unlock(&lock);
+    // UNLOCK
+    // **********
+
+    return result;
+}
+
+int pdo::enclave_api::block_store::BlockStorePut(
+    const uint8_t* key,
+    const size_t keySize,
+    const uint8_t* value,
+    const size_t valueSize
+    )
+{
+    std::string keyStr = BinaryToHexString(key, keySize);
+    std::string valueStr = BinaryToHexString(value, valueSize);
+
+    Log(PDO_LOG_DEBUG, "Block Store Put: %zu bytes '%s' -> %zu bytes '%s'",
+        keySize, keyStr.c_str(), valueSize, valueStr.c_str());
+    // **********
+    // LOCK
+    pthread_spin_lock(&lock);
+
+    map[keyStr] = valueStr;
+
+    pthread_spin_unlock(&lock);
+    // UNLOCK
+    // **********
+
+    return PDO_SUCCESS;
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+int pdo::enclave_api::block_store::BlockStoreHead(
+    const ByteArray& inKey
+    )
+{
+    uint8_t *value;
+    size_t value_size;
+
+    // Fetch the state from the block storage
+    int ret = BlockStoreGet(inKey.data(), inKey.size(),
+                            &value, &value_size);
+    if (ret != 0) {
+        // No data found - return -1 for size
+        return -1;
+    }
+
+    // Found data - return its size
+    return (int)value_size;
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+pdo_err_t pdo::enclave_api::block_store::BlockStoreGet(
+    const ByteArray& inKey,
+    ByteArray& outValue
+    )
+{
+    pdo_err_t result = PDO_SUCCESS;
+
+    uint8_t *value;
+    size_t value_size;
+
+    // Fetch the state from the block storage
+    int ret = BlockStoreGet(inKey.data(), inKey.size(),
+                            &value, &value_size);
+    pdo::error::ThrowIf<pdo::error::ValueError>(
+       ret != 0, "Unable to get from Block Store");
+
+    // Copy the buffer back to the caller's ByteArray
+    outValue.resize(value_size);
+    outValue.assign(value, value + value_size);
+
+    return result;
+}
+
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+pdo_err_t pdo::enclave_api::block_store::BlockStorePut(
+    const ByteArray& inKey,
+    const ByteArray& inValue
+    )
+{
+    pdo_err_t result = PDO_SUCCESS;
+
+    int ret = BlockStorePut(inKey.data(), inKey.size(),
+                            inValue.data(), inValue.size());
+
+    pdo::error::ThrowIf<pdo::error::ValueError>(
+       ret != 0, "Unable to put into the Block Store");
```

Comment:
this is really the wrong kind of error to be throw at this point. Use ThrowPDOError

Reply:
Fixed in upcoming commit.